applications/MeshingApplication/custom_utilities/tet10_refinement_utility.h

```diff
@@ -0,0 +1,499 @@
+	bool create_element = false;
+	int number_elem = 0;
+	int splitted_edges = 0;
+	int internal_node = 0;
+
+	const ProcessInfo& rCurrentProcessInfo = this_model_part.GetProcessInfo();
+	int edge_ids[6];
+	int t[56];
+	std::vector<int> aux;
+
+        std::cout << "****************** REFINING MESH ******************" << std::endl;
+        std::cout << "OLD NUMBER ELEMENTS: " << rElements.size() << std::endl;
+
+	for (int & i : t)
+	{
+	    i = -1;
+	}
+
+	for (ElementsArrayType::iterator it = it_begin; it != it_end; ++it)
+	{
+	  Element::GeometryType& geom = it->GetGeometry();
+          CalculateEdges(geom, Coord, edge_ids, aux);
+
+	  create_element = TetrahedraSplit::Split_Tetrahedra(edge_ids, t, &number_elem, &splitted_edges, &internal_node);
+
+	  if (create_element == true)
+	  {
+	      to_be_deleted++;
+
+          GlobalPointersVector< Element >& rChildElements = it->GetValue(NEIGHBOUR_ELEMENTS);
+          // We will use this flag to identify the element later, when operating on
+          // SubModelParts. Note that fully refined elements already have this flag set
+          // to true, but this is not the case for partially refined element, so we set it here.
+          it->SetValue(SPLIT_ELEMENT,true);
+          rChildElements.resize(0);
+	      
+		  unsigned int i0 = aux[0];
+		  unsigned int i1 = aux[1];
+		  unsigned int i2 = aux[2];
+		  unsigned int i3 = aux[3];
+          unsigned int i4 = aux[4];
+		  unsigned int i6 = aux[5];
+		  unsigned int i7 = aux[6];
+		  unsigned int i5 = aux[7];
+          unsigned int i8 = aux[8];
+		  unsigned int i9 = aux[9];
+          
+
+		  Tetrahedra3D10<Node < 3 > > geom(
+		      this_model_part.Nodes()(i0),
+		      this_model_part.Nodes()(i1),
+		      this_model_part.Nodes()(i2),
+		      this_model_part.Nodes()(i3),
+              this_model_part.Nodes()(i4),
+		      this_model_part.Nodes()(i5),
+		      this_model_part.Nodes()(i6),
+		      this_model_part.Nodes()(i7),
+              this_model_part.Nodes()(i8),
+		      this_model_part.Nodes()(i9)
+		  );
+
+		  // Generate new element by cloning the base one
+		  Element::Pointer p_element;
+          const Element& rElem = KratosComponents<Element>::Get("Element3D10N");
+		  p_element = rElem.Create(current_id, geom, it->pGetProperties());
+		  p_element->Initialize(rCurrentProcessInfo);
+		  p_element->InitializeSolutionStep(rCurrentProcessInfo);
+		  p_element->FinalizeSolutionStep(rCurrentProcessInfo);
+
+		  // Setting the internal variables in the child elem
+		  if (interpolate_internal_variables == true)
+		  {
+		      InterpolateInteralVariables(number_elem, *it.base(), p_element, rCurrentProcessInfo);
+		  }
+
+		  // Transfer elemental variables
+		  p_element->Data() = it->Data();
+		  p_element->GetValue(SPLIT_ELEMENT) = false;
+		  NewElements.push_back(p_element);
+
+          rChildElements.push_back( Element::WeakPointer(p_element) );
+
+		  current_id++;
+	      it->SetId(large_id);
```

Comment:
This mechanism for deleting can have side effect in other modelparts sharing the elements. I would suggest to use the flag (TO_REMOVE) and standard modelpart erase mechanism

@rubenzorrilla and @RiccardoRossi this is used in the base class also. IMO, to be replaced there also.

Reply:
Totally agree. Thanks for the suggestion, already updated it in this code.
(PD: i think you meant the TO_ERASE flag 😄)